File: LeetCode/LeetCode/reverse_list.cc
#include<iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL) {}
};

class Solution{
    public:
        ListNode *reverseBetween(ListNode *head, int m, int n) {
            ListNode dummy{0};
            dummy.next = head;

            auto *prev = &dummy;

            for(int i = 0; i < m - 1; ++i) {
                prev = prev -> next;
            }

            auto *head2 = prev;
            prev = prev -> next;
            auto *cur = prev->next;

            for(int i = m; i < n; ++i) {
                prev->next = cur->next;
            }

            for(int i = m; i < n; ++i) {
                prev->next = cur->next;
                cur->next = head2->next;
                head2->next = cur;
                cur = prev->next;
            }

            return dummy.next;
        }
};

int main(int argc, char ** argv)
{
    int i, count = argc - 3;

    ListNode dummy(-1);
    ListNode *prev = & dummy;
    ListNode *p;

    for (i = 0; i < count; ++i) {
        p = new ListNode(atoi(argv[i+3]));
        prev->next = p;
        prev = p;
    }

    int m = atoi(argv[1]);
    int n = atoi(argv[2]);
    struct ListNode *head = Solution().reverseBetween(dummy.next, m, n);
    for(p = head; p!=NULL; p = p->next)
    {
        printf("%d", p->val);
    }
    printf("\n");
    return 0;
};

File: LeetCode/LeetCode/reverse_list.go
package main

import "fmt"

type ListNode struct {
	Val int
	Next *ListNode
}

func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if m == n {
		return head
	}
	headPre := &ListNode{}
	headPre.Next = head
	m++
	n++

	mPre, mNode, nNext := split(headPre, m ,n)
	h, e := reverse(mNode)
	mPre.Next = h
	e.Next = nNext
	return headPre.Next

}

func split(head *ListNode, m,n int) (mPre, mNode, nNext *ListNode) {
	i := 1
	for head != nil {
		if i == m -1 {
			mPre = head
			mNode = head.Next
		}

		if i == n {
			nNext = head.Next
			head.Next = nil
			break
		}

		head = head.Next
		i++
	}

	return
}

func reverse(head *ListNode)(h, e *ListNode) {
	if head == nil || head.Next == nil {
		return head, head
	}

	var end *ListNode

	h, end = reverse(head.Next)
	end.Next = head
	e = head
	return
}

func main() {
	head := &ListNode{
		1,
		&ListNode{
			2,
			&ListNode{
				3, 
				&ListNode{
					4,
					&ListNode{
						5,
						nil,
					},
				},
			},
		},
	}
	reverseBetween(head, 2, 4)
	for item:=head; item!=nil; {
		fmt.Println(item.Val)
		item = item.Next
	}
}
